### ISSUE TITLE
Grandpa Benchmarks are not usable as is

### ISSUE THREAD

**Issue by ETeissonniere** [paritytech/substrate#7383]
Using a command of the type `benchmark --chain dummy --raw --wasm-execution compiled --execution wasm --pallet pallet_grandpa --extrinsic "*" --steps 50 --repeat 2 --output` would generate a file similar to:
```rust
use frame_support::weights::{Weight, constants::RocksDbWeight as DbWeight};

pub struct WeightInfo;
impl pallet_grandpa::WeightInfo for WeightInfo {
	// WARNING! Some components were not used: ["x"]
	fn check_equivocation_proof() -> Weight {
		(135_500_000 as Weight)
	}
	fn note_stalled() -> Weight {
		(6_000_000 as Weight)
			.saturating_add(DbWeight::get().writes(1 as Weight))
	}
}

```

However, it cannot be plugged as is into the `WeightInfo` value of `pallet_grandpa::Trait` which isn't compliant with `pallet_grandpa::WeightInfo` as defined [here](https://github.com/paritytech/substrate/blob/master/frame/grandpa/src/lib.rs#L105).

**Comment by bkchr** [comment]
CC @shawntabrizi @andresilva

**Comment by shawntabrizi** [comment]
@ETeissonniere Grandpa does not have good benchmarks for the end to end automation.

You can see the manual weight file we generated here: https://github.com/paritytech/substrate/blob/master/frame/grandpa/src/benchmarking.rs

@andresilva will know better what it will take (if possible) to better automate these weights, but for now, you can likely rely on this implementation as "good enough".

**Comment by andresilva** [comment]
Yeah as @shawntabrizi mentioned we don't have a contained benchmark for `report_equivocation`, instead we have benchmarks for all the methods it calls out to and we manually derive the weight from there. The issue is that in order to benchmark `report_equivocation` you need to setup on-chain state properly in order to:
- Have a proper validator staking setup to equivocate;
- Be able to sign statements with those validators' keys in order to produce valid equivocations.

I honestly don't remember exactly what I got stuck on but I wasn't able to do this setup when I wrote those benchmarks.

**Comment by ETeissonniere** [comment]
Got it, is there a way to reproduce the same computing by hand so that we can fine tune the weights ourselves? Or shall we just stick to the default?

**Comment by andresilva** [comment]
Yes, you should run the following benchmarks:
- `session::check_membership_proof_current_session`
- `grandpa::check_equivocation_proof`
- `offences::report_offence_grandpa`

These intermediary weights are reflected here: https://github.com/paritytech/substrate/blob/master/frame/grandpa/src/default_weights.rs#L35

**Comment by Janislav** [comment]
I'm facing the same problem right now, but I'm not really sure if I understand the discussed solution here. As far as my understanding goes I've to do the following things:

1. Run the benchmark tool for grandpa against my runtime
2. Extend the resulting struct with the missed function
3. In this function I combine the resulting benchmarks of
    - session::check_membership_proof_current_session
    - grandpa::check_equivocation_proof
    - offences::report_offence_grandpa

Is this ruffly right? Would be awesome to get some confirmation here. Or is it easier and I just have to copy what's in the function of `default_weight.rs` of the grandpa crate?
